node/src/Pos/Block/Network/Retrieval.hs

```diff
@@ -305,17 +306,20 @@ handleCHsValid enqueue endedRecoveryVar nodeId lcaChild newestHash = do
                 -- difficulty than ncrecoveryheader, we're
                 -- gracefully exiting recovery mode.
                 let isMoreDifficultThan b x = b ^. difficultyL >= x ^. difficultyL
-                endedRecovery <- atomically $ do
+                atomically $ do
                     mRecHeader <- tryReadTMVar recHeaderVar
                     case mRecHeader of
                         Nothing -> return False
                         Just (_, rHeader) ->
                             if any (`isMoreDifficultThan` rHeader) blocks
                                 then isJust <$> tryTakeTMVar recHeaderVar
                                 else return False
-                void $ tryPutMVar endedRecoveryVar endedRecovery
-    void $ waitForConversations its
+    results <- waitForConversations $ fmap try convs
```

Comment:
> We're creating a single conversation here, so it will have a single result, right?

Not necessarily. The outbound queue is ultimately in control of how many times the conversation body is run (how many peers are reached). In practice it will be either 1 or 0 because the set in `MsgRequestBlocks` is a singleton.

`results :: Map NodeId (Either SomeException Bool))`. For each peer that was actually contacted, the result of the conversation (whether recovery mode was ended) or an exception if the conversation did not complete normally. The fold using `Any` tells us whether any of these conversations gave `Right True`, i.e. whether recovery mode ended.